prov/verbs/src/verbs_msg_ep.c

```diff
@@ -561,30 +560,36 @@ static inline int fi_ibv_poll_reap_unsig_cq(struct fi_ibv_msg_ep *ep)
 	return FI_SUCCESS;
 }
 
-/* WR must be filled out by now except for context */
 static inline ssize_t
-fi_ibv_send(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr, void *context)
+fi_ibv_send_poll_cq_if_needed(struct fi_ibv_msg_ep *ep, struct ibv_send_wr *wr)
 {
-	struct fi_ibv_wre *wre = NULL;
 	int ret;
 
+try_again:
+	ret = FI_IBV_INVOKE_POST(send, send, ep->id->qp, wr, NULL);
+	if (OFI_UNLIKELY(ret == -FI_EAGAIN)) {
+		ret = fi_ibv_poll_reap_unsig_cq(ep);
+		if (OFI_UNLIKELY(ret))
+			return -FI_EAGAIN;
+		goto try_again;
```

Comment:
I think this ends up looping until a prior send completes, so the current send can be posted.

Reply:
Did you mean change the following:
```
        ret = FI_IBV_INVOKE_POST(send, send, ep->id->qp, wr, NULL);
	if (OFI_UNLIKELY(ret == -FI_EAGAIN)) {
		ret = fi_ibv_poll_reap_unsig_cq(ep);
		if (OFI_UNLIKELY(ret))
			return -FI_EAGAIN;
		ret = FI_IBV_INVOKE_POST(send, send, ep->id->qp, wr, NULL);
```
Actually, I didn't get your idea

Reply:
When can fi_ibv_poll_reap_unsign_cq() return 0?  Does this always indicate that it has found a completion, or can 0 indicate that nothing was read from the CQ?

Reply:
> When can fi_ibv_poll_reap_unsign_cq() return 0?

when `ibv_poll_cq` returns >= 0.



> Does this always indicate that it has found a completion

Returning 0 indicates that function call was successful (this doesn't say anything about "were there any entries read from CQ?"). It means that `ibv_poll_cq` doesn't return an error.

Reply:
The issue is that if the send queue is full, this change basically loops reading the cq until a completion is read and the send can be posted.  If the work request at the head of the send queue is large, this could result in a significant delay.  Rather than looping over posting-reading the cq, we should only attempt to poll the cq once and retry the posting only if we found a completion.  So, something like:
```
ret = post_send(..);
if (ret == EAGAIN) {
    ret = poll_cq(..);
    if (ret > 0)
        ret = post_send(..);
    else if (ret == 0)
        ret = EAGAIN;
}
```
with the code actually dealing with whatever return codes are correct.